R/token-create.R

```diff
@@ -0,0 +1,135 @@
+#' Create a terminal token
+#'
+#' Creates a terminal token represented as (a row of) a parse table.
+#' @param tokens Character vector with tokens to create.
+#' @param texts Character vector with texts of the token to create.
+#' @param lag_newlines Character vector with lag_newlines corresponding to the
+#'   tokens.
+#' @param spaces Character vector with spaces corresponding to the tokens.
+#' @param pos_ids Character vector with positional id corresponding to the
+#'   tokens.
+#' @param parents Vector with `id` corresponding to the
+#'   parent of the tokens we want to create.
+#' @param token_before Character vector corresponding to the columns
+#'   `token_before`.
+#' @param token_after Character vector corresponding to the columns
+#'   `token_after`.
+#' @param indention_ref_ids Character vector with indention ref ids
+#'   corresponding to the tokens.
+#' @param indents Vector with indents corresponding to the tokens.
+#' @family token creators
+create_tokens <- function(tokens,
+                          texts,
+                          lag_newlines = 0,
+                          spaces = 0,
+                          pos_ids,
+                          parents,
+                          token_before = NA,
+                          token_after = NA,
+                          indention_ref_ids = NA,
+                          indents = 0) {
+  len_text <- length(text)
+  data_frame(
+    token = tokens,
+    text  = texts,
+    short = substr(texts, 1, 5),
+    lag_newlines = lag_newlines,
+    newlines = lead(lag_newlines),
+    pos_id = pos_ids,
+    parent = parents,
+    token_before = token_before,
+    token_after = token_after,
+    id = NA,
+    terminal = rep(TRUE, len_text),
+    internal = rep(FALSE, len_text),
+    spaces = spaces,
+    multi_line = rep(FALSE, len_text),
+    indention_ref_id = indention_ref_ids,
+    line1 = rep(NA, len_text),
```

Comment:
I'd rather omit NA columns, and let `bind_rows()` take care of it.

Reply:
Our `bind_rows()` implementation does only support matching column names.
``` r
styler:::bind_rows(data.frame(a = 1, b = 3), data.frame(b = 3, c = 4))
#> Error in match.names(clabs, names(xi)): names do not match previous names
```

Reply:
So maybe just rever that to the initial PR.